### PR TITLE
Fix Orders in confirmation state cannot be completed by hub or shop

### PR DESCRIPTION
**Pull request by cyrillefr**
#### What? Why?

- addresses (in part) #11188 

#### What should we test?
Cf. #11188 ('Steps to reproduce' part)

This is only a start to see what is left to be done.

### CODE REVIEW

**Comment by mkllnk** [comment]
Is this ready for review?

**Comment by cyrillefr** [comment]
Hello, @mkllnk ,

Sorry, I did not kown if I had to add some label or not.
Yes, it can be reviewed.

**Comment by cyrillefr** [comment]
Hello @mkllnk , 

I have submitted this PR to show the first part of #11180 , (see discussion there, I am looking for insights :smiley: ).

Since it does not resolves the #11188 issue, I guess it is too early for reviewing.

**Comment by mkllnk** [comment]
That's good to know, @cyrillefr. The PR title contains `[WIP]` and therefore I wasn't sure if you were still working on it or wanted feedback.

You code change may be part of the solution. Changing the state machine is tricky and can have other side effects. It will need thorough testing. And and you mentioned, there's more work to do to resolve the issue. So I'll move this back to In Dev but I reckon that you are on the right path.

**Comment by cyrillefr** [comment]
Hello, I have just committed some new code.
I will try to explain what I did.

I replaced the original capture event by a new one which is capture + complete order.
It was not possible to add a complete action if the original button was clicked, so I made a new event.
I kept the original styling.

I tracked (by settings breakpoints) what is done when one is to click the "complete order" as a client and
what is done by the code I have added(order.complete): the same breakpoints were hit.

I also noticed that it was possible to capture on the list of orders in the admin dashboard.
And it is not implemented the same way: the first payment of the order is captured when the icon is clicked.
I did not modify this also because I do not know if what I did in the payment part(replacing an event with an other)
will be the accepted solution.

**Review by rioug — COMMENTED** [review]
This looks good :clap: . I had play with it and I believe it fix the issue, to be confirmed by the testers

File: app/controllers/spree/admin/payments_controller.rb [hunk 1]
```diff
@@ -140,7 +140,8 @@ def load_data
       #
       # Otherwise redirect user to that step
       def can_transition_to_payment
-        return if @order.payment? || @order.complete? || @order.canceled? || @order.resumed?
+        return if @order.confirmation? || @order.payment? ||
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/controllers/spree/admin/payments_controller.rb:143]: I could imagine a partial payment triggering this. :shrug:

**Review by mkllnk — APPROVED** [review]
Looking good. :+1:

**Review by rioug — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @cyrillefr ,

Thanks so much for sticking with this one!

First, I've reproduced the issue again, on current master: indeed, orders in the confirmation state cannot be completed by the hub, as the payment section is not accessible:

image

After this PR, the Payment section becomes accessible :+1:  

image

This looks great... I've splitted the test by payment methods.

**Stripe**
For orders left in the confirmation state and paid with Stripe, we can observe an inconsistency, as the capture button is still displays, although it cannot be manually captured in the BO by the hub; If the capture button is clicked (I think this would be the first thing a hub manager would try to do) then the payment fails, and other payments can be added on top of this one - the pic below shows a cash payment, added by the hub:

image

This would be ok to live with I think. The main issue can be seen though, when we check the Stripe dashboard:

image

Unfortunately, when the order moves to `complete` and if the card is valid, the payment is captured on Stripe... This, however, is not showing up on the order... So, for this particular order we have 2 x payments in the amount of `8.41 $` a cash and the Stripe payment.

This is a tricky situation I think, which could lead us to double charging of customers. I don't think we can merge this just yet.

So, the fix works very nicely @cyrillefr, but unfortunately only for cash and Paypal: 

**Paypal**
The payment fails here but charging does not occur as the customer does not log in to their account; this allows adding a cash payment to the order and brings it to a completed state:
image
 
**Cash**
This fix works perfectly for orders left in the confirmation state, which were initiated it cash; accessing the payments tab allows the payment to be captured manually, without any issues:
image

We need a bit of though around this one... I'll leave it with the `feedback needed` label. What say you @openfoodfoundation/train-drivers-product-owners ?

**Comment by filipefurtad0** [comment]
Ahh, so, it seems the behavior actually depends on the type of card used... which makes sense.

Considering the [Stripe test cards](https://stripe.com/docs/testing#authentication-and-setup):

- for cards which always require authentication like `x-3184` -> the payment fails indeed (after clicking capture) and it is safe to add a cash payment, leading to completion
- it is with cards like `x-3155` which authentication is only asked when this is set up (the so called `Authenticate unless set up` card) that the payment might be captured, when the order is completed.

Seems a bit of an edge case... Bit, if there are double charging it surely will be painful to trace down...

**Comment by filipefurtad0** [comment]
Scratch that :-D

The [pic here ](https://user-images.githubusercontent.com/49817236/283566306-5c20680e-25f8-4f52-b03c-d41f233d4c3b.png)refers to another order, another payment - my bad. The info never gets sent to Stripe, so no payment intent is created. We're good to go on the payments side of things :+1:

**Pull request merged**